easybuild/easyblocks/r/raxml.py

```diff
@@ -0,0 +1,194 @@
+RAXML_INSTRUCT_SETS = ["AVX2", "AVX", "SSE3"]
+RAXML_CPU_FEATURES = {
+    "SSE3": ["sse3", "see4_1", "sse4_2"],
+    "AVX": ["avx", "avx1.0"],
+    "AVX2": ["avx2"],
+}
+# Supported parallelization features grouped in non-MPI and MPI
+RAXML_PARALLEL_FEATURES = {
+    "nompi": [None, "PTHREADS"],
+    "mpi": ["MPI", "HYBRID"],
+}
+
+
+class EB_RAxML(MakeCp):
+    """Support for building and installing RAxML."""
+
+    @staticmethod
+    def extra_options(extra_vars=None):
+        """Change default values of options"""
+        extra = MakeCp.extra_options()
+        # files_to_copy is not mandatory
+        extra["files_to_copy"][2] = CUSTOM
+        return extra
+
+    def __init__(self, *args, **kwargs):
+        """RAxML easyblock constructor, define class variables."""
+        super(EB_RAxML, self).__init__(*args, **kwargs)
+
+        def filter_optarch_features(support_features):
+            """
+            Filter list of supported CPU instruction sets based on optarch string
+            """
+            if not isinstance(support_features, list):
+                support_features = list(support_features)
+            unfilter_features = [feat.upper() for feat in support_features]
+
+            optarch = build_option("optarch") or self.toolchain.options.get("optarch", "")
+            # optarch is a dictionary with settings per compiler family
+            if isinstance(optarch, dict):
+                comp_fam = self.toolchain.comp_family()
+                optarch = optarch.get(comp_fam, "")
+            # convert boolean optarch settings to string equivalents
+            if optarch is False:
+                optarch = OPTARCH_GENERIC
+            elif optarch is True:
+                optarch = ""
+            optarch = optarch.upper()
+
+            # check for generic build
+            if optarch == OPTARCH_GENERIC:
+                self.log.debug("Building generic RAxML per optarch setting: %s", optarch)
+                return [None]
+            # check for supported labels in optarch (by given priority order)
+            for instruct_set in unfilter_features:
+                if instruct_set in optarch:
+                    dbg_msg = "Restricting supported features in RAxML to '%s' per optarch: %s"
+                    self.log.debug(dbg_msg, instruct_set, optarch)
+                    return [instruct_set]
+            # anything else is unsupported
+            if optarch:
+                print_warning("Unsupported 'optarch' configuration setting for RAxML, ignoring: %s" % optarch)
+
+            return support_features
+
+        def has_cpu_feature(feature):
+            """
+            Check if host CPU architecture has given feature
+            """
+            if HAVE_ARCHSPEC:
+                import archspec.cpu
+
+                host = archspec.cpu.host()
+                return feature.lower() in host
+
+            try:
+                return any(f in get_cpu_features() for f in RAXML_CPU_FEATURES[feature])
```

Comment:
Would be nicer with a single `return` as the end, I'm not a big fan of inline returns, see also https://realpython.com/python-return-statement/#remembering-the-return-value (but that's partially personal, I guess)

Reply:
Maybe this function should be moved into framework, it's pretty generic... Same for `filter_optarch_features`